odbc_fdw.c

```diff
@@ -1645,163 +1702,150 @@ odbcIterateForeignScan(ForeignScanState *node)
 		/* Loop through the columns */
 		for (i = 1; i <= columns; i++)
 		{
-			SQLLEN indicator;
-			char * buf;
-
 			int mask_index = i - 1;
 			int col_size = list_nth_int(col_size_array, mask_index);
 			int mapped_pos = list_nth_int(col_position_mask, mask_index);
 			ColumnConversion conversion = list_nth_int(col_conversion_array, mask_index);
+			SQLSMALLINT target_type = SQL_C_CHAR;
+			SQLLEN result_size;
+			int chunk_size, effective_chunk_size;
+			int buffer_size = 0;
+			char * buffer = 0;
+			char * hex;
+			int used_buffer_size = 0;
+			GetDataTruncation truncation;
+			bool binary_data = false;
+			if (conversion == BIN_CONVERSION)
+			{
+				target_type	= SQL_C_BINARY;
+				binary_data = true;
+			}
 
-			if (col_size == 0) {
+			if (col_size == 0)
+			{
 				col_size = 1024;
 			}
 
+			chunk_size = binary_data ? col_size : col_size + 1;
+
 			/* Ignore this column if position is marked as invalid */
 			if (mapped_pos == -1)
 				continue;
 
-			buf = (char *) palloc(sizeof(char) * (col_size+1));
-
-			/* retrieve column data as a zero-terminated string */
-			/* TODO:
-			   binary fields (SQL_C_BIT, SQL_C_BINARY) do not have
-			   a trailing zero; they should be copied as now but without
-			   adding 1 to col_size, or using SQL_C_BIT or SQL_C_BINARY
-			   and then encoded into a binary PG literal (e.g. X'...'
-			   or B'...')
-			   For floating point types we should use SQL_C_FLOAT/SQL_C_DOUBLE
-			   to avoid precision loss.
-			   For date/time/timestamp these structures can be used:
-			   SQL_C_TYPE_DATE/SQL_C_TYPE_TIME/SQL_C_TYPE_TIMESTAMP.
-			   And finally, SQL_C_NUMERIC and SQL_C_GUID could also be used.
-			*/
-			buf[0] = 0;
-			ret = SQLGetData(stmt, i, SQL_C_CHAR,
-			                 buf, sizeof(char) * (col_size+1), &indicator);
-
-			if (ret == SQL_SUCCESS_WITH_INFO)
+			while (1) // Reading loop for reading the field in chunks
```

Comment:
Scary and unclear `while(1)`. Would something like the following be more clear?
```c
truncation = STRING_TRUNCATION;
while (truncation == STRING_TRUNCATION)
{
...
}
```

And based on this, maybe it would make more sense to do a
```c
do
{

} while (truncation == STRING_TRUNCATION) + other condition
```

Reply:
Nice suggestion, looks better now (I'm still leaving a early exit of the loop that could probably be handled more elegantly, but may brain's too tired now to pursue it).